### PR TITLE
Implicit token def quick fix

### PR DESCRIPTION
**Pull request by wojciszek**
Provide quick fix for missing token definition from fragments.

### CODE REVIEW

**Comment by bjansen** [comment]
I haven't forgotten you PR, I just need to find some time to look at it with a well rested brain :)

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    private void appendTokenDefAtLastLine(@Nullable("is null when called from inspection") Editor editor) {
+        String tokenName = element.getReference().getCanonicalText();
+        int lastLineOffset = editor.getDocument().getLineEndOffset(editor.getDocument().getLineCount() - 1);
+        editor.getDocument().insertString(lastLineOffset, "\n" + buildTokenDefinitionText(tokenName));
+    }
+
+    @NotNull
+    static String buildTokenDefinitionText(String tokenName) {
+        String tokenDefRule = tokenName.toUpperCase().chars().mapToObj(c -> (char) c).map(String::valueOf).collect(Collectors.joining(" "));
```
Comment (bjansen) [inline src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: I'm not convinced this is a sensible default value. What about `MYTOKEN: 'MYTOKEN'` with the value automatically selected in the editor to make it easy to modify it?

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        super.visitElement(element);
+    }
+
+    private void handleLeafElement(LexerRuleRefNode element, ProblemsHolder holder) {
+        if (!(element.getElementType() instanceof TokenIElementType)) {
+            return;
+        }
+
+        TokenIElementType tokenIElementType = (TokenIElementType) element.getElementType();
+        if (!"TOKEN_REF".equalsIgnoreCase(org.antlr.intellij.plugin.parser.ANTLRv4Lexer.VOCABULARY.getDisplayName(tokenIElementType.getANTLRTokenType()))) {
```
Comment (bjansen) [inline src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: I believe you can use `ANTLRv4TokenTypes` to avoid the string comparison:
```java
if (element.getElementType() != ANTLRv4TokenTypes.TOKEN_ELEMENT_TYPES.get(ANTLRv4Lexer.TOKEN_REF))
```

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            return;
+        }
+
+        for (GrammarIssue grammarIssue : grammarIssues) {
+            boolean isAmongOffendingTokens = isAmongOffendingTokens(element, grammarIssue);
+            if (isAmongOffendingTokens) {
+                Optional<LocalQuickFix> localQuickFix = getFix(element, grammarIssue.getMsg().getErrorType());
+
+                if (localQuickFix.isPresent()) {
```
Comment (bjansen) [inline src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: The warning is registered twice (once by `ANTLRv4ExternalAnnotator`, and once by your code):

image

One possible solution would be to modify `ANTRLv4Inspection` to run only when `isOnTheFly` is `true`, and let `ANTLRv4ExternalAnnotator` register the quick fix. Another benefit is that `GrammarIssuesCollector.collectGrammarIssues()` will be called once instead of twice when the current editor is modified.

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java (hunk 1 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: I can add such default as well. I've chosen `MYTOKEN: M Y T O K E N;` because it allows us to create token based on fragments (instead of literal) which is I think better when it comes to case-insensitive languages.

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java (hunk 3 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: I was thinking about replacing ANTLRv4ExternalAnnotator completely with ANTRLv4Inspection. As far as I'm concerned Annotators does not allow to bind quick fix to code. What do you think?

Comment (bjansen) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: Actually you can register quick fixes if you retrieve the result of `createXxxAnnotation()`:

```java
Annotation errorAnnotation = holder.createErrorAnnotation(range, issue.annotation);
errorAnnotation.registerFix(...);
```

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java (hunk 1 above)
Comment (bjansen) [reply src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: OK if you think it's better then go for it. As long as there is the automatic selection, I'm happy :)

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java (hunk 3 above)
Comment (bjansen) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: I couldn't remember the big differences between inspections and annotators, but found a nice explanation [on this page](https://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/code_inspections_and_intentions.html?search=inspec). 

Inspections can be disabled, which means that if we replace the existing annotator with a single inspection, then disabling it will disable the whole highlighting.

On the other hand, creating multiple inspections will probably provide poor performance because the problems are computed by the ANTLR tool. We can't possibly call the tool once per inspection.

I'm not sure running your newly created quick fix in batch mode is relevant, because the user will probably have to edit each automatic fix manually to replace things like `M Y T O K E N`.

To summarize: it seems to me that the current annotator is the correct way to highlight all these problems because they come from a single source (the output of the ANTLR tool). We should probably modify it to register quick fixes when they exist.

We can still add inspections later on for other kinds of problems for which an automatic fix does not require further editing.

WDYT?

Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: After reading your explanation i think that i will try to stick with annotator and remove inspection.

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java (hunk 2 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: Code removed.

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java (hunk 1 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: I'm having problem with automatic selection. I've managed to create selection such as rename selection (is when you press enter the name is accepted). Do you happen to know how to implement this correctly? 

My selection:
my_token_select

Selection that I want to implement:
rename_select

File: src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java (hunk 3 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java]: I have removed inspections and registered fix with annotation.

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java (hunk 1 above)
Comment (bjansen) [reply src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: This could be done using an inline refactoring, but in our case `M Y T O K E N` is not a a `PsiNamedElement` so that won't work. I managed to do a similar thing in Ceylon IDE for IntelliJ [using `TemplateManagerImpl`](https://github.com/eclipse/ceylon-ide-intellij/blob/1f228e03a3eac5643c0928c90613c9203cd28c00/source/org/eclipse/ceylon/ide/intellij/platform/IdeaLinkedMode.ceylon#L95-L106).

Basically you need to:
* create a `TemplateBuilderImpl` on your `PsiFile`
* call `builder.replaceRange(range, new TextExpression("M Y L O O K U P"))` for each text range you want to edit
* call `TemplateManager.getInstance(project).startTemplate(editor, template)` to start the template

By doing this you will basically transform the whole file into a live template with a single editable variable :)

**Comment by bjansen** [comment]
Also, could you please add your name to contributors.txt?

File: src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java (hunk 1 above)
Comment (wojciszek) [reply src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java]: Done. I managed to do this without making whole file a template - looked up on ceylon repo :)

**Comment by bjansen** [comment]
Travis says you're using a method which is not available in IntelliJ SDK <= 2018.2:
```
/home/travis/build/antlr/intellij-plugin-v4/src/main/java/org/antlr/intellij/plugin/validation/AddTokenDefinitionFix.java:92: error: cannot find symbol
        return new TextRange(tokenExprTextOffset, elementAt.getTextRangeInParent().getEndOffset() - 1);
```

Could you fix this please?

**Comment by bjansen** [comment]
Thanks again for all your work!
